aiohttp/client.py

```diff
@@ -986,61 +1016,112 @@ def _prepare_headers(self, headers: Optional[LooseHeaders]) -> "CIMultiDict[str]
                     added_names.add(key)
         return result
 
-    def get(
-        self, url: StrOrURL, *, allow_redirects: bool = True, **kwargs: Any
-    ) -> "_RequestContextManager":
-        """Perform HTTP GET request."""
-        return _RequestContextManager(
-            self._request(hdrs.METH_GET, url, allow_redirects=allow_redirects, **kwargs)
-        )
+    if sys.version_info >= (3, 11) and TYPE_CHECKING:
+        import typing
```

Comment:
MyPy weirdly doesn't like a direct import of `Unpack` from `typing`, but is fine with `typing.Unpack`.

Reply:
I don't see any errors when I test it locally?

Reply:
Let me show you an example:

Reply:
The same example I posted above (as is now):

```bash
mypy --version
mypy 1.10.0 (compiled: yes)
```

```bash
mypy aiohttp/test.py --python-version=3.11

Success: no issues found in 1 source file
```

Using `from typing import Unpack`

Now:

```python
    if sys.version_info >= (3, 11) and TYPE_CHECKING:
        from typing import Unpack

        def get(
            self,
            url: StrOrURL,
            **kwargs: Unpack[_RequestOptions],
        ) -> "_RequestContextManager": ...

        def options(
            self,
            url: StrOrURL,
            **kwargs: Unpack[_RequestOptions],
        ) -> "_RequestContextManager": ...

        def head(
            self,
            url: StrOrURL,
            **kwargs: Unpack[_RequestOptions],
        ) -> "_RequestContextManager": ...

        def post(
            self,
            url: StrOrURL,
            **kwargs: Unpack[_RequestOptions],
        ) -> "_RequestContextManager": ...

        def put(
            self,
            url: StrOrURL,
            **kwargs: Unpack[_RequestOptions],
        ) -> "_RequestContextManager": ...

        def patch(
            self,
            url: StrOrURL,
            **kwargs: Unpack[_RequestOptions],
        ) -> "_RequestContextManager": ...

        def delete(
            self,
            url: StrOrURL,
            **kwargs: Unpack[_RequestOptions],
        ) -> "_RequestContextManager": ...
```


```bash
aiohttp/client.py:1025:23: error: Variable "aiohttp.client.ClientSession.Unpack" is not valid as a type  [valid-type]
                **kwargs: Unpack[_RequestOptions],
                          ^
aiohttp/client.py:1025:23: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiohttp/client.py:1025:23: note: See https://mypy.rtfd.io/en/stable/_refs.html#code-valid-type for more info
aiohttp/client.py:1031:23: error: Variable "aiohttp.client.ClientSession.Unpack" is not valid as a type  [valid-type]
                **kwargs: Unpack[_RequestOptions],
                          ^
aiohttp/client.py:1031:23: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiohttp/client.py:1037:23: error: Variable "aiohttp.client.ClientSession.Unpack" is not valid as a type  [valid-type]
                **kwargs: Unpack[_RequestOptions],
                          ^
aiohttp/client.py:1037:23: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiohttp/client.py:1043:23: error: Variable "aiohttp.client.ClientSession.Unpack" is not valid as a type  [valid-type]
                **kwargs: Unpack[_RequestOptions],
                          ^
aiohttp/client.py:1043:23: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiohttp/client.py:1049:23: error: Variable "aiohttp.client.ClientSession.Unpack" is not valid as a type  [valid-type]
                **kwargs: Unpack[_RequestOptions],
                          ^
aiohttp/client.py:1049:23: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiohttp/client.py:1055:23: error: Variable "aiohttp.client.ClientSession.Unpack" is not valid as a type  [valid-type]
                **kwargs: Unpack[_RequestOptions],
                          ^
aiohttp/client.py:1055:23: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiohttp/client.py:1061:23: error: Variable "aiohttp.client.ClientSession.Unpack" is not valid as a type  [valid-type]
                **kwargs: Unpack[_RequestOptions],
                          ^
aiohttp/client.py:1061:23: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
aiohttp/test.py:10:25: note: Revealed type is "def (url: Union[builtins.str, yarl.URL], **kwargs: Unpack?[TypedDict('aiohttp.client._RequestOptions', {'params'?: Union[typing.Mapping[builtins.str, builtins.str], None], 'data'?: Any, 'json'?: Any, 'cookies'?: Union[typing.Mapping[builtins.str, Union[builtins.str, http.cookies.BaseCookie[builtins.str], http.cookies.Morsel[Any]]], typing.Iterable[tuple[builtins.str, Union[builtins.str, http.cookies.BaseCookie[builtins.str], http.cookies.Morsel[Any]]]], http.cookies.BaseCookie[builtins.str], None], 'headers'?: Union[typing.Mapping[Union[builtins.str, multidict.istr], builtins.str], multidict.CIMultiDict[builtins.str], multidict.CIMultiDictProxy[builtins.str], None], 'skip_auto_headers'?: Union[typing.Iterable[builtins.str], None], 'auth'?: Union[tuple[Any, Any, Any, fallback=aiohttp.helpers.BasicAuth], None], 'allow_redirects'?: builtins.bool, 'max_redirects'?: builtins.int, 'compress'?: Union[builtins.str, None], 'chunked'?: Union[builtins.bool, None], 'expect100'?: builtins.bool, 'raise_for_status'?: Union[None, builtins.bool, def (aiohttp.client_reqrep.ClientResponse) -> typing.Awaitable[None]], 'read_until_eof'?: builtins.bool, 'proxy'?: Union[builtins.str, yarl.URL, None], 'proxy_auth'?: Union[tuple[Any, Any, Any, fallback=aiohttp.helpers.BasicAuth], None], 'timeout'?: Union[aiohttp.client.ClientTimeout, aiohttp.helpers._SENTINEL, None], 'ssl'?: Union[ssl.SSLContext, builtins.bool, aiohttp.client_reqrep.Fingerprint], 'server_hostname'?: Union[builtins.str, None], 'proxy_headers'?: Union[typing.Mapping[Union[builtins.str, multidict.istr], builtins.str], multidict.CIMultiDict[builtins.str], multidict.CIMultiDictProxy[builtins.str], None], 'trace_request_ctx'?: Union[types.SimpleNamespace, None], 'read_bufsize'?: Union[builtins.int, None], 'auto_decompress'?: Union[builtins.bool, None], 'max_line_size'?: Union[builtins.int, None], 'max_field_size'?: Union[builtins.int, None]})]) -> aiohttp.client._RequestContextManager"
Found 7 errors in 1 file (checked 1 source file)
```

Reply:
The revealed type is still fine, but it causes MyPy issues in the source file.

Reply:
Right, I missed the python-version. That error makes sense though, because you're creating a `ClientSession.Unpack` class attribute. I think it's best to move the import to the top of the module and just have 2 if conditions.

Reply:
Fixed.

Reply:
https://github.com/aio-libs/aiohttp/pull/8463/commits/e4af6b13c7e4d7ebc5dcac5a7d38bdf6a13619fc